src/storage/src/hummock/iterator/forward_user.rs

```diff
@@ -102,15 +103,14 @@ impl<I: HummockIterator<Direction = Forward>> UserIterator<I> {
             }
 
             if &self.last_key.user_key.as_ref() != key {
-                self.last_key.set(full_key);
+                self.last_key = full_key.to_vec().into_bytes();
                 // handle delete operation
                 match self.iterator.value() {
                     HummockValue::Put(val) => {
                         if self.delete_range_aggregator.should_delete(key, epoch) {
                             self.stats.skip_delete_key_count += 1;
                         } else {
-                            self.last_val.clear();
-                            self.last_val.extend_from_slice(val);
+                            self.last_val = Bytes::copy_from_slice(val);
```

Comment:
`to_vec` and `copy_from_slice` will cause memory allocation.

I suspect the extra memory allocation may cause non-negligible overhead. Can we run a micro-bench to see how bad it can be? Is it possible to avoid using `Bytes` in user iterator?

Reply:
In current main, though we don't have memory allocation in the user iterator, the value returned from `IterStream` of `iter` method is created by `Bytes::copy_from_slice`, so the memory allocation is unavoidable. Same for the key.

Besides, the `Bytes::copy_from_slice` involves one memory allocation and one memory copy, plus the memory copy caused by `Vec::extend_from_slice` in user iterator, there are one memory allocation and two memory copies when looked end to end. In this PR, there is only one memory allocation and one memory copy, so with this PR, there is no extra overhead introduced, and we even avoid one more memory copy.

Reply:
> In current main, though we don't have memory allocation in the user iterator, the value returned from `IterStream` of `iter` method is created by `Bytes::copy_from_slice`, so the memory allocation is unavoidable. Same for the key.
> 
> Besides, the `Bytes::copy_from_slice` involves one memory allocation and one memory copy, plus the memory copy caused by `Vec::extend_from_slice` in user iterator, there are one memory allocation and two memory copies when looked end to end. In this PR, there is only one memory allocation and one memory copy, so with this PR, there is no extra overhead introduced, and we even avoid one more memory copy.

That's great. Thanks for the explanation.

Reply:
> So why there is no extra overhead?

When only looking at the code in the user iterator, there is extra overhead introduced. However, when we look from end to end, the stream returned from `StateStoreRead::iter` emits value of static `Bytes`, which is created by `Bytes::copy_from_slice` and involves one memory allocation and copy. Plus the memory copy in `extend_from_slice`, there is one memory allocation and two copies.

Reply:
Thanks! I get it now. We should change

https://github.com/risingwavelabs/risingwave/blob/8d7af5b7f193fd89f2c427cf7ed4932f878bce4f/src/storage/src/hummock/store/state_store.rs#L302

to `clone`.